x-pack/solutions/security/plugins/security_solution/public/flyout/document_details/right/components/highlighted_fields_modal.tsx

```diff
@@ -0,0 +1,270 @@
+  { defaultMessage: 'Add custom' }
+);
+
+const SELECT_PLACEHOLDER = i18n.translate(
+  'xpack.securitySolution.flyout.right.investigation.highlightedFields.modalSelectPlaceholder',
+  { defaultMessage: 'Select or search for options' }
+);
+
+const CommonUseField = getUseField({ component: Field });
+
+interface InvestigationFieldsFormData {
+  customHighlightedFields: string[];
+}
+
+const schema: FormSchema<InvestigationFieldsFormData> = {
+  customHighlightedFields: {
+    fieldsToValidateOnChange: ['customHighlightedFields'],
+    type: FIELD_TYPES.COMBO_BOX,
+    validations: [{ validator: fieldValidators.emptyField('error') }],
+  },
+};
+
+interface HighlighedFieldsModalProps {
+  /**
+   * Available fields to select from
+   */
+  fieldOptions: DataViewFieldBase[];
+  /**
+   * The data formatted for field browser
+   */
+  dataFormattedForFieldBrowser: TimelineEventsDetailsItem[];
+  /**
+   * The rule
+   */
+  rule: RuleResponse;
+  /**
+   * The custom highlighted fields
+   */
+  customHighlightedFields: string[];
+  /**
+   * The function to set the edit loading state
+   */
+  setIsEditLoading: (isEditLoading: boolean) => void;
+  /**
+   * The function to set the modal visible state
+   */
+  setIsModalVisible: (isModalVisible: boolean) => void;
+}
+
+/**
+ * Modal for editing the highlighted fields of a rule.
+ */
+export const HighlighedFieldsModal: FC<HighlighedFieldsModalProps> = ({
+  fieldOptions,
+  rule,
+  customHighlightedFields,
+  dataFormattedForFieldBrowser,
+  setIsEditLoading,
+  setIsModalVisible,
+}) => {
+  const { addSuccess } = useAppToasts();
+  const { euiTheme } = useEuiTheme();
+  const { mutateAsync: updateRule } = useUpdateRule();
+  const modalTitleId = useGeneratedHtmlId();
+
+  const defaultFields = useHighlightedFields({
+    dataFormattedForFieldBrowser,
+    investigationFields: customHighlightedFields,
+    type: 'default',
+  });
+  const defaultFieldsArray = useMemo(() => Object.keys(defaultFields), [defaultFields]);
+
+  const customFields = useMemo(
+    () => customHighlightedFields.map((field: string) => ({ label: field })),
+    [customHighlightedFields]
+  );
+
+  const [selectedOptions, setSelectedOptions] = useState(customFields);
+
+  const options = useMemo(
+    () => fieldOptions.map((field) => ({ label: field.name })),
+    [fieldOptions]
+  );
+
+  const { form } = useForm({
+    defaultValue: { customHighlightedFields: [] },
+    schema,
+  });
+
+  const onCancel = useCallback(() => {
+    setIsModalVisible(false);
+  }, [setIsModalVisible]);
+
+  const onSubmit = useCallback(async () => {
+    setIsEditLoading(true);
+
+    const updatedRule = await updateRule({
+      ...rule,
+      id: undefined,
+      investigation_fields:
+        selectedOptions.length > 0
+          ? { field_names: selectedOptions.map((option) => option.label) }
+          : undefined,
+    } as RuleUpdateProps);
+
+    addSuccess(SUCCESSFULLY_SAVED_RULE(updatedRule?.name ?? 'rule'));
```

Comment:
Shouldn't we handle a failure here?

Reply:
failure is handled in the hook https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/security_solution/public/detection_engine/rule_management/logic/use_update_rule.ts#L12